Nullptr crash in WebCore::HTMLAnchorElement::isKeyboardFocusable

pratiksha-99 opened the pull request:
#### 952a0bb866457207750ab33fe8e35aa6bd101007
<pre>
Nullptr crash in WebCore::HTMLAnchorElement::isKeyboardFocusable
<a href="https://bugs.webkit.org/show_bug.cgi?id=281282">https://bugs.webkit.org/show_bug.cgi?id=281282</a>
<a href="https://rdar.apple.com/137177639">rdar://137177639</a>

Reviewed by Ryosuke Niwa.

Added nullptr check for the LocalFrame before focused keyboard events.
This does not have a reliable test case but have tested locally on Mac
that WebKit does not have this crash.

* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::isKeyboardFocusable const):
* Source/WebCore/svg/SVGAElement.cpp:
(WebCore::SVGAElement::isKeyboardFocusable const):

Canonical link: <a href="https://commits.webkit.org/285145@main">https://commits.webkit.org/285145@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/295d27b361e97d555c9314367c62cef7281c8730

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71682 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51095 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24456 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75796 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22887 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73797 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58896 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22707 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15093 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74748 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46355 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61745 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43016 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19212 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21228 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64921 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19575 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77516 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15916 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18762 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64326 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15960 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61778 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64327 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15843 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12482 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6117 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46895 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/1674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47966 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49250 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47708 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/fast/events/focus-keyboard-event-crash.html:
```diff
@@ -0,0 +1,30 @@
+<script>
+(async () => {
+    globalThis.testRunner?.waitUntilDone();
```
Please also call:
```
globalThis.testRunner?.dumpAsText();
```

rniwa commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -32,6 +32,7 @@ fast/events/ios/ipad [ Skip ]
 fast/forms/ios [ Pass ]
 fast/forms/ios/ipad [ Skip ]
 fast/forms/switch [ Pass ]
+fast/events/focus-keyboard-event-crash.html [ Pass ]
```
No, you need to say either `Skip` or `WontFix`. Also, this isn't the right place to insert this expectation. Move to more towards the end of the file.

pratiksha-99 replied:
Updated.

rniwa commented on LayoutTests/platform/mac-wk1/TestExpectations:
```diff
@@ -6,6 +6,7 @@
 #//////////////////////////////////////////////////////////////////////////////////////////
 
 fast/forms/attributed-strings.html [ Pass ]
+fast/events/focus-keyboard-event-crash.html [ Pass ]
```
Ditto.

rniwa commented on LayoutTests/fast/events/focus-keyboard-event-crash.html:
```diff
@@ -0,0 +1,28 @@
+    } catch {}
+
+    dialog.showModal();
+    globalThis.testRunner?.notifyDone();
```
Before this line, can we add something like `document.write('This test passes WebKit does not crash<br>PASS');` right before notifyDone so that there is some output text?
After that, you need to run the test with `run-webkit-tests --reset <path-to-test>` to update the expected result.

pratiksha-99 replied:
I added this locally to test but the expected test file is not getting updated.

pratiksha-99 committed:
Nullptr crash in WebCore::HTMLAnchorElement::isKeyboardFocusable
https://bugs.webkit.org/show_bug.cgi?id=281282
rdar://137177639

Reviewed by Ryosuke Niwa.

Added nullptr check for the LocalFrame before focused keyboard events.
This does not have a reliable test case but have tested locally on Mac
that WebKit does not have this crash.

* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::isKeyboardFocusable const):
* Source/WebCore/svg/SVGAElement.cpp:
(WebCore::SVGAElement::isKeyboardFocusable const):

Canonical link: https://commits.webkit.org/285145@main
